lib/experimental/graph.nim

```diff
@@ -0,0 +1,428 @@
+## Generic implementation of graph data structure
+
+import std/[
+  tables,
+  strformat,
+  deques,
+  hashes,
+  options,
+  math,
+  lenientops,
+  strutils,
+  sequtils,
+  algorithm,
+  intsets,
+  sets
+]
+
+#[
+checklist
+
+- [ ] All implementations work both at compile and run-time
+- [ ] Works on JS target
+- [ ] All iterator algorithms have both mutable and immutable implementations
+- [ ] As little side effects as possible
+- [ ] Make node and edge ID's distinct integers to avoid messing them up
+- [ ] Iterateo over node value pairs (source, target: N, edge: E)
+
+- IDEA :: Add `->` and `<-` as operators that would return `TempEdge`
+  object that can be added via `graph.add source -> target`.
+
+- TODO :: Test if cycle detection works for graphs that contain overlapping
+  cycles (`0 -> 1 -> 0` and `1 -> 2 -> 1`)
+
+]#
+
+type
+  HGraphProperty* = enum
+    gpDirected
+    gpUndirected
+    gpAllowSelfLoops
+    gpIgnoreSelfLoops
```

Comment:
```suggestion
    gpSelfLoopsAllowed
    gpSelfLoopsIgnored
```

Not strongly suggesting this, but it doesn't keep the tense parallel with the other options.